MoltenVK/MoltenVK/Commands/MVKCmdTransfer.mm

```diff
@@ -74,12 +80,42 @@ static inline MTLSize mvkClampMTLSize(MTLSize size, MTLOrigin origin, MTLSize ma
             return cmdBuff->reportError(VK_ERROR_FEATURE_NOT_PRESENT, "vkCmdCopyImage(): Cannot copy between incompatible formats, such as formats of different pixel sizes, or between images with different sample counts.");
         }
         
-		_vkImageCopies.push_back(vkIR);
+        _vkImageCopies.emplace_back(std::move(vkIR2));
```

Comment:
I feel like building a full `VkCopyImageInfo2` on the stack wouldn't make sense, because the number of regions is variable and we therefore can't allocate the VkImageCopy2 on the stack and need a call to `new` or `malloc`, which is wasteful, realistically. So instead, I would opt for something that merges the two together by adding an inline function that does the validation that can be called from both functions. That would look something like this:
```cpp
template <size_t N>
VkResult MVKCmdCopyImage<N>::setContent(MVKCommandBuffer* cmdBuff,
                                        VkImage srcImage,
                                        VkImageLayout srcImageLayout,
                                        VkImage dstImage,
                                        VkImageLayout dstImageLayout,
                                        uint32_t regionCount,
                                        const VkImageCopy* pRegions) {
    _srcImage = (MVKImage*)srcImage;
    _srcLayout = srcImageLayout;

    _dstImage = (MVKImage*)dstImage;
    _dstLayout = dstImageLayout;

    _vkImageCopies.clear();		// Clear for reuse
    for (uint32_t regionIdx = 0; regionIdx < regionCount; regionIdx++) {
        auto& vkIR = pRegions[regionIdx];
        VkImageCopy2 vkIR2 = {
            VK_STRUCTURE_TYPE_IMAGE_COPY_2, nullptr,
            vkIR.srcSubresource, vkIR.srcOffset,
            vkIR.dstSubresource, vkIR.dstOffset,
            vkIR.extent
        };
        validateRegion(&vkIR2);
        _vkImageCopies.emplace_back(std::move(vkIR2));
    }
    
    return VK_SUCCESS;
}

template <size_t N>
VkResult MVKCmdCopyImage<N>::setContent(MVKCommandBuffer* cmdBuff,
                                        const VkCopyImageInfo2* pCopyImageInfo) {
    _srcImage = (MVKImage*)pCopyImageInfo->srcImage;
    _srcLayout = pCopyImageInfo->srcImageLayout;

    _dstImage = (MVKImage*)pCopyImageInfo->dstImage;
    _dstLayout = pCopyImageInfo->dstImageLayout;

    _vkImageCopies.clear();        // Clear for reuse
    _vkImageCopies.reserve(pCopyImageInfo->regionCount);
    for (uint32_t regionIdx = 0; regionIdx < pCopyImageInfo->regionCount; regionIdx++) {
        auto& vkIR = pCopyImageInfo->pRegions[regionIdx];
        validateRegion(&vkIR);
        _vkImageCopies.push_back(vkIR);
    }

    return VK_SUCCESS;
}

VkResult MVKCmdCopyImage<N>::validateRegion(VkImageCopy2* region) {
    uint8_t srcPlaneIndex = MVKImage::getPlaneFromVkImageAspectFlags(region->srcSubresource.aspectMask);
    uint8_t dstPlaneIndex = MVKImage::getPlaneFromVkImageAspectFlags(region->dstSubresource.aspectMask);

    // Validate
    MVKPixelFormats* pixFmts = cmdBuff->getPixelFormats();
    if ((_dstImage->getSampleCount() != _srcImage->getSampleCount()) ||
        (pixFmts->getBytesPerBlock(_dstImage->getMTLPixelFormat(dstPlaneIndex)) != pixFmts->getBytesPerBlock(_srcImage->getMTLPixelFormat(srcPlaneIndex)))) {
        return cmdBuff->reportError(VK_ERROR_FEATURE_NOT_PRESENT, "vkCmdCopyImage(): Cannot copy between incompatible formats, such as formats of different pixel sizes, or between images with different sample counts.");
    }
    return VK_SUCCESS;
}
```

Reply:
> because the number of regions is variable and we therefore can't allocate the VkImageCopy2 on the stack  and need a call to `new` or `malloc`, which is wasteful, realistically

Yeah. We definitely want to stay away from heap allocations. But since these POD structs, I was talking about using a C-style dynamically-sized array on the stack. Something like:

    VkImageCopy2 regions2[regionCount];
    VkCopyImageInfo2 copyImageInfo2;
    copyImageInfo2.pRegions = regions2;
    for (uint32_t regionIdx = 0; regionIdx < regionCount; regionIdx++) {
        auto& vkIR = pRegions[regionIdx];
        regions2[regionIdx] = {
            VK_STRUCTURE_TYPE_IMAGE_COPY_2, nullptr,
            vkIR.srcSubresource, vkIR.srcOffset,
            vkIR.dstSubresource, vkIR.dstOffset,
            vkIR.extent
        };
    ...   // populate the rest of copyImageInfo2

Instead of the first line, you could alternately use the following, but `N` will be larger than `regionCount`, so it consumes more on the stack than strictly necessary.

    MVKSmallVector< VkImageCopy2, N> regions2;

Reply:
Well, at that point I would just extract the validation into a separate function. The two paths are simply different and so I would just keep them different. Also, VLAs are not supported by default in C++ so I would very much like to not use them.

Reply:
We use VLA's for C-syle POD structs in MoltenVK already. But I'm fine if you don't want to use them.

So just refactor what you can to reduce code duplication for best practices, and resubmit the PR, please and thanks.